쥬스 메이커 [STEP 2] paul, rieul

jwonyLee opened the pull request:
안녕하세요 토니! @Monsteel 

폴(@ParkHyeongSeok), 리을(@jwonyLee)입니다.
`STEP 2` 완료해서 PR 요청 드립니다. ㅎㅎ 이번에도 잘 부탁 드려요!

## 작업 내용

- 과일 제조 화면 → 과일 재고 화면 이동 구현
  - 화면 표시 방식을 Modal을 사용한 이유
    - HIG의 Modality 항목을 살펴 보면, `현재 작업과 다른 작업을 선택하거나 사람들의 주의를 집중시키는 것이 중요한 경우에만 모달을 사용하라.` 라는 내용이 있습니다. 또한 Navigation의 경우 계층적인 구조를 표현할 때 사용하는데, 재고 관리는 계층 구조보다 독립적인 작업에 가깝다고 여겨져 Modal을 사용해서 표시하게 되었습니다.
  - 코드를 이용해 화면을 전환하고자 했으며, 이를 위해 뷰 컨트롤러 인스턴스를 간편하게 만들기 위해  `StoryboardBindable` 프로토콜을 만들고, `UpdateStockViewController`가 채택하게 했습니다. 이 과정에서 불가피하게 강제 형변환을 사용하게 되었습니다.
- 과일 재고 화면 → 과일 제조 화면 이동 구현
  - 과일 재고 화면을 `fullScreen`으로 표시하면서 과일 제조 화면으로 돌아갈 수 있는 방법이 없어 불가피하게 `STEP 3`에서 구현해야 하는 `닫기` 기능을 먼저 구현했습니다. 😢
- 과일 재고 표시
  - 과일 재고를 표시하기 위해서 `JuiceMakerViewController`가 `FruitStore`의 인스턴스를 갖게 했습니다. 그리고 화면이 표시되기 전(`viewWillAppear`)에 재고 정보를 가져와서 이후에 구현할 `UpdateStockViewController`에서 데이터가 변경되어도 화면이 다시 표시되면서 갱신할 수 있도록 구현했습니다.
- 주스 주문 기능 구현
  - 모든 주문 버튼에 `@IBAction` 메서드를 만드는 것은 비효율적이라고 생각했습니다.
  - 그래서 모든 주스 주문 버튼에 `tag`를 부여했습니다. 그리고 `Juice` 타입을 `Int`를 채택하게 해서 `rawValue`로 각 `tag`와 매칭할 수 있게 했습니다. 
  - 주스 주문 버튼들은 `orderJuice(_:)`를 호출하고, `Juice` 타입을 만들어서 `makeJuice(_:)`를 호출합니다. `makeJuice(_:)`로부터 에러를 받거나, 어떤 주스가 나왔는 지 메세지를 받을 수 있도록 수정했습니다.
    - 재고가 모자라지 않고 주스를 만들 수 있으면 과일 재고를 다시 그리고, 어떤 주스가 나왔는 지 받은 메세지를 알럿으로 표시했습니다.
    - 재고가 모자라면 실패 알럿을 표시했습니다.
  - 주스 제조 후 결과로 주스 이름을 포함하여 반환하기 위해 `Juice`에 computed property로 `name`을 구현했습니다.

## 코드를 작성하면서 고민되었던 점

- `makeJuice()`의 리턴이 `String`인 경우는 모호하고, `Juice`인 경우 결국 파라미터로 받은 `juice`를 그대로 리턴하게 되는데, 새로운 타입을 만드는게 더 좋은 결정인지에 대한 고민이 있습니다:)
- `JuiceMakerViewController`는 과일의 재고를 표시하기 위해서 `FruitStore`가 필요하고, 주스를 제조하기 위해서 `JuiceMaker`가 필요합니다. `JuiceMaker`는 주스를 만들기 위해 `FruitStore`가 필요한 형태이고요. 과일 재고와 관련해서 `JuiceMakerViewController`와 `STEP 3`에서 구현하게 될 `UpdateStockViewController`, `JuiceMaker` 모두 같은 `FruitStore` 인스턴스를 가리키고 있어야 합니다. 이를 위해 현재 `JuiceMakerViewController`에 있는 `JuiceMaker`는 `lazy` 키워드를 이용해 `JuiceMakerViewController.fruitStore`를 주입해서 생성하고 있습니다. 이러한 상황에서 `FruitStore`를 싱글톤으로 만드는 것이 좋을 지, 아니면 현재 방식처럼 `UpdateStockViewController`를 표시할 때도 `JuiceMakerViewController`에서 `fruitStore`를 주입하는 것이 좋을 지 고민입니다.

## 해결이 되지 않은 점

## 조언을 얻고 싶은 부분

- UI 작업은 main thread 에서 해야 하는 걸로 알고 있는데요! main thread에 태우는(?) 것을 해당 메서드를 호출하는 시점에서 하는 것이 좋은지, 메서드 내부에서 하는 것이 좋은지 궁금합니다.
  ```swift
  DispatchQueue.main.async {
      self.configureUI()
  }

  func configureUI() {
      // UI 작업
  }
  ```
  ```swift
  self.configureUI()

  func configureUI() {
      DispatchQueue.main.async {
          // UI 작업
      }
  }
  ```
- 이번에 `STEP 2`를 진행하면서 과일 재고를 업데이트하는 메서드 `updateFruitStock()`을 main thread로 감쌌는데 반드시 필요한 작업일까요? 간혹 Xcode 에서 UI 작업이 main thread 에서 실행되지 않고 있을 때 경고를 띄워주는데, `updateFruitStock()`은 해당 경고가 뜨지 않더라구요. UI 작업은 무조건적으로 main thread 에서 실행하는 게 좋은 지 궁금합니다. :(
- `JuiceMakerViewController`에 `FruitStore`와 `JuiceMaker`에 관한 의존성을 주입해야 하는지 고민이 됩니다. `JuiceMakerViewController`를 외부 객체와 독립적으로 운영하게 좋다고 생각하는데, 한편 ViewController에서 의존성을 주입하는 것을 눈으로 많이 보지 못해서 꺼려지기도 합니다. 일반적인 패턴인지, 한다면 어떤 의존성 주입 패턴(이니셜라이저, 프로퍼티, 함수 등)을 사용하는 것이 좀 더 자연스러울지 궁금해서 여쭙습니다:)

ParkHyeongSeok committed:
feat: 화면 이동 구현

- JuiceMakerViewController의 presentUpdateStockScene 버튼 구현
- 폴더 구조 수정
- StoryboardBindable 프로토콜 제작

jwonyLee committed:
feat: 뒤로가기 기능 구현

- 재고 수정 화면에서 주스 제조 화면으로 돌아가는 기능 구현

jwonyLee committed:
feat: 과일 재고 표시

jwonyLee committed:
feat: 주스 주문 기능 추가

- JuiceMaker를 이용해 주스 제조
- 주스가 완성되면 완성 알럿 표시
- 재고가 부족하면 재고 부족 알럿 표시

jwonyLee committed:
feat: 주스 제조 후 결과로 주스 이름을 포함하여 반환

- Juice 에 computed property 로 name 구현

Monsteel commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,92 @@
+        let updateStockViewController: UpdateStockViewController = .instantiate()
+        updateStockViewController.modalPresentationStyle = .fullScreen
+        updateStockViewController.modalTransitionStyle = .coverVertical
+        let navigationController: UINavigationController = UINavigationController(rootViewController: updateStockViewController)
```
`UINavigationController`를 사용하신 이유가 궁금합니다! 🙋🏻‍♂️
`navigationBar` 를 사용하기 위함이라면, 다른 방법도 있을것 같아요 👀

Monsteel commented on JuiceMaker/JuiceMaker/Controller/JuiceMakerViewController.swift:
```diff
@@ -0,0 +1,92 @@
+    
+    private func showFailAlert() {
+        let alert: UIAlertController = UIAlertController(title: "제조 실패",
+                                                         message: "재료가 모자라요. 재고를 수정할까요?",
```
재료가 모자라서 생긴 오류인지 확신 할 수 없는 상황인것 같아요  🤔

Monsteel reviewed (changes requested):
안녕하세요 폴 리을~!
Step2 진행하시느라 수고 많으셨습니다 🎉

남겨주신 코멘트들과 코드를 보면서 폴과 리을이 코드에 대해 심도있게 고민하고, 더 나은 방향을 진심으로 고려하시는 모습이 엿보여서 저도 덩달아 가슴이 웅장해졌습니다 ☀️
아주 잘 해주시고 계세요 !

특히 `StoryboardBindable`을 통해 코드에서의 불편한 Storyboard연결을 간단하게 할 수 있게 해주신 점이 정말 인상깊었습니다 👍

관련해서 궁금한점도 하나 있는데요~!
Button의 Segue를 통해 다음 스토리보드를 열수도 있었을것 같은데 그 방법을 선택하지 않으시고 이런 구현방법을 선택하신 이유가 궁금합니다 🙋🏻‍♂️

그리고 이어서 남겨주신 질문에 대한 답변을 달아볼게요!

> `makeJuice()`의 리턴이 `String`인 경우는 모호하고, `Juice`인 경우 결국 파라미터로 받은 `juice`를 그대로 리턴하게 되는데, 새로운 타입을 만드는게 더 좋은 결정인지에 대한 고민이 있습니다:)

Juice를 받아 Juice를 반환하는 것에 대해 저는 전혀 이질감이 없다고 생각합니다.
프로그래밍 세계이기 때문에 가능한게 아닌가 싶어요.

들어오는 주문은 `Order`로 묶고 나가는 음료수는 `Product`로 구분했다고 해도,
`Order`와 `Product`의 구현은 ~~거의~~완전히 일치하게 되는 케이스의 구현이 나올텐데, 프로그래밍 세계라서 오히려 `Juice`구현으로 통합할 수 있는게 아닌가 싶습니다.

`프로그래밍 세계는` 현실세계를 메타포 하지만, 현실 세계를 100% 프로그래밍 세계에 반영할 필요는 없다고 생각합니다.
프로그래밍 세계와 현실세계의 차이를 장점으로 이용하는 프로그래밍적 사고를 저는 중요하게 생각합니다 🙋🏻‍♂️

> UI 작업은 main thread 에서 해야 하는 걸로 알고 있는데요! main thread에 태우는(?) 것을 해당 메서드를 호출하는 시점에서 하는 것이 좋은지, 메서드 내부에서 하는 것이 좋은지 궁금합니다.
>  ```swift
>  DispatchQueue.main.async {
>      self.configureUI()
>  }
>
>  func configureUI() {
>      // UI 작업
>  }
>  ```
>  ```swift
>  self.configureUI()
>
>  func configureUI() {
>      DispatchQueue.main.async {
>          // UI 작업
>      }
>  }
>  ```

음 해당 부분에 대해서는 저는 개인적으로 후자의 방법을 선호하는데요,
전자의 방법의 경우 `configureUI()`을 사용할 때 마다 매번 `DispatchQueue.main.async { ... }`를 감싸야 할지에 대한 판단을 해야하고, 만약 사용해야 하는 경우라면 매번 감싸는 형태로 가져가야 하기 때문입니다.

다만, `DispatchQueue.main.async { ... }`를 사용하지 않아도 되는 상황에서 `DispatchQueue.main.async { ... }`를 감싼 형태의 코드가 실행되는것이 싫다!
고 하시면 전자도 좋은 방법이 될 수 있을것 같습니다.

저의 이런 판단 기준이 선택에 조금이나마 도움이 될 수 있기를 바랍니다 🙏
하지만 이번 Step에서는 해당 코드가 필요하지 않을것 같아요. 자세한건 다음 질문에서 이야기해 볼게요 !

> 이번에 `STEP 2`를 진행하면서 과일 재고를 업데이트하는 메서드 `updateFruitStock()`을 main thread로 감쌌는데 반드시 필요한 작업일까요? 간혹 Xcode 에서 UI 작업이 main thread 에서 실행되지 않고 있을 때 경고를 띄워주는데, `updateFruitStock()`은 해당 경고가 뜨지 않더라구요. UI 작업은 무조건적으로 main thread 에서 실행하는 게 좋은 지 궁금합니다. :(

해당 작업이 이미 main thread에서 작동했기 때문이에요.
당장은 간단한 연산이라서 딜레이가 체감이 안될 수 있지만,
이 프로젝트에 시간이 오래걸리는 연산작업이 진행된다면 UI까지 멈추는 현상이 발생할거에요. 👍

예를들어
``` swift
 var value = 0
 for i in 0...9999999999999999999999 { value += i }
```
이런 코드를 중간에 넣으셔서 확인 가능합니다

![스크린샷 2022-02-18 오후 7 18 36](https://user-images.githubusercontent.com/52942409/154663618-ee81aa2d-b28b-415d-8097-0786532716c9.png)

의존성에 대한 질문을 주셨는데, 한꺼번에 대답을 드릴 수 있을것 같아요

해당 프로젝트에서 가장 좋은 방법은 최상위 level에서 의존성을 관리하면서
각 ViewController에 주입해주는 방향이 제일 좋을것 같습니다 🤔
-> 자연스럽게 화면 전환하는 코드가 상위로 올라가게 될 것 같네요

이 부분에 대해서는 제 힌트(?)를 가지고서 두분이서 고민해 보시면 좋을것 같아요.
만약 이해하시는데에 어려움이 있다면 언제든 저에게 말씀주시면 재차 공유드릴 수 있도록 하겠습니다! 🤔

두 의존성 관련 질문에 대답이 되었으면 좋겠습니다 🙏

코드 리뷰는 간단한 것들 몇가지만 남겼습니다~! 확인해 주시면 감사하겠습니다 🙇🏻‍♂️
추가로 궁금한 점이 하나 있는데, 프로젝트의 파일 구조를 API / SupportingFiles / Service 등... 과 같이 나누신 기준이 궁금합니다 🙋🏻‍♂️

jwonyLee committed:
refactor: main thread 제거

jwonyLee committed:
refactor: UpdateStockViewController NavigationBar로 변경

- NavigationController 에서 NavigationBar로 변경

jwonyLee committed:
refactor: JuiceMakerViewController.orderJuice 리팩토링

- JuiceMakerError 를 명확하게 나누어서 처리
- JuiceMaker.makeJuice 에서 String 반환을 Juice로 변경
- 성공한 경우 Juice를 전달해서, 성공 메시지를 JuiceMakerViewController에서 처리
- 실패한 경우 Error를 전달해서, 에러 메시지를 JuiceMakerError.localizedDescription으로 표시

jwonyLee committed:
test: hasStock 초기재고에서 키위의 재고가 7개 이상 있다

- 사용하지 않는 값 `_ =` 처리
- default: 0 을 옵셔널 처리

jwonyLee committed:
refactor: StoreType 추가

- 의존성 제거를 위한 StoreType 프로토콜 추가

ParkHyeongSeok committed:
refactor: Juice와 JuiceMaker 간접화

- MakerType 프로토콜 제작 : JuiceMaker 간접화
- Drinkable 프로토콜 제작 : Juice 간접화
- StoreType 제거 : FruitStore는 구상클래스로 유지

ParkHyeongSeok committed:
refactor: SceneDelegate에서 JuiceMaker 및 FruitStore 의존성 주입

- Storyboard 없이 런칭되도록 수정

jwonyLee commented:
안녕하세요 토니!  @Monsteel 
코드 리뷰해주신 내용 잘 읽어 봤습니다. 예시까지 들어주셔서 감사해요! 

> Button의 Segue를 통해 다음 스토리보드를 열수도 있었을것 같은데 그 방법을 선택하지 않으시고 이런 구현방법을 선택하신 이유가 궁금합니다 🙋🏻‍♂️

이 질문에 대한 답변을 해볼게요! `음료 제조 화면`에서 `과일 재고 화면`으로 이동하는 경우는 두 가지가 있는데요.
1. 우측 상단 `재고 수정`버튼을 통해 이동하는 경우
2. 주스를 제조할 때 재고가 부족해서 이동하는 경우

두 가지 경우를 대응할 때 코드로 이동을 구현하면 이동하는 코드를 재사용할 수 있다는 장점이 있어서 해당 방식을 선택하게 되었어요.

## 작업 내용

코드 리뷰 내용 반영과 함께 의존성의 관한 부분을 수정했어요.

- NavigationBar 사용으로 변경
  - 저랑 폴 둘 다, 네비게이션 바 사용은 익숙하지 않아서 `UINavigationController`를 사용했었는데, 코드 리뷰에 남겨주신 것처럼 `UINavigationBar`를 직접 사용하는 것이 더 나은 것 같아요!
  - `UINavigationBar`는 스토리보드에서 작업했고, 하단에 나타나는 줄을 제거하는 부분은 코드로 작성했어요.
- 재료가 모자라서 생긴 오류인지 확신할 수 없는 상황 수정
  - 기존 코드는 모든 에러 처리를 동일하게 했었는데요, 저희가 정의해놓은 오류가 단일 오류(`.outOfStock`)이어서 별도의 처리가 필요없다고 생각했었어요.
  - 오류를 조금 더 명확하게 처리하기 위해서, `JuiceMakerError.outOfStock`을 별도로 catch 하고, 그 외의 알 수 없는 오류에 대한 케이스를(`.unknown`)을 별도로 추가했어요.
  - `LocalizedError`를 채택, 구현해서 에러 메시지를 가지도록 하고, `showFailAlert(error:)`에서 해당 에러에 대한 메시지를 표시하도록 수정했어요.
- `JuiceMaker`는 `MakerType`, `Juice`는 `Drinkable` 프로토콜을 만들어서 추상화에 의존하도록 수정했어요.
  - 저희가 처음부터 추상화를 기반으로 설계를 시작했으면 더 수월했을텐데, 중간 단계부터 도입하려고 하니까 어렵더라구요. 😢 이번 기회에 하나 배우고 갑니다.. `FruitStore`와 `Fruit`까지 추상화 시켰으면 좋았겠지만, 난도가 있어서 해결하지 못했고, 할 수 있는 부분부터 해결했습니다.
  - 이 과정에서 `SceneDelegate`에서 JuiceMaker, FruitStore 의존성을 주입하게끔 수정했어요.

Monsteel reviewed (approved):
폴 리을 Step2도 수고많으셨습니다~! 🎉
